### PR TITLE
Subobjects and layers support

### PR DESCRIPTION
**Pull request by pkratten**

### CODE REVIEW

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 1]
```diff
@@ -221,6 +277,10 @@ public Rhino.Geometry.Mesh[] GetMeshes(RhinoObject rhinoObject)
                     mesh.EnsurePrivateCopy();
                     validMeshes.Add(mesh);
                 }
+                else
+                {
+                    validMeshes.Add(null);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:282]: Why are we adding a null mesh to the array here?

File: glTF-BinExporter/RhinoMaterialGltfConverter.cs [hunk 2]
```diff
@@ -39,7 +39,7 @@ public int AddMaterial()
             // Prep
             glTFLoader.Schema.Material material = new glTFLoader.Schema.Material()
             {
-                Name = rhinoMaterial.Name,
+                Name = string.IsNullOrEmpty(renderMaterial.Name) ? renderMaterial.Id.ToString() : renderMaterial.Name,
```
Comment (jrz371) [inline glTF-BinExporter/RhinoMaterialGltfConverter.cs:42]: Can you make this name resolving it's own method?

File: glTF-BinExporter/RhinoDocGltfConverter.cs [hunk 3]
```diff
@@ -118,43 +163,54 @@ public byte[] GetBinaryBuffer()
...
 
-            if(!materialsMap.TryGetValue(materialId, out int materialIndex))
-            {
-                RhinoMaterialGltfConverter materialConverter = new RhinoMaterialGltfConverter(options, binary, dummy, binaryBuffer, material, workflow);
-                materialIndex = materialConverter.AddMaterial();
-                materialsMap.Add(materialId, materialIndex);
+                if (material == null && options.UseDisplayColorForUnsetMaterials)
+                {
+                    Color4f objectColor = GetObjectColor(rhinoObject);
+                    materialIndices[i] = CreateSolidColorMaterial(objectColor, string.IsNullOrEmpty(rhinoObject.Name) ? rhinoObject.Id.ToString() : rhinoObject.Name);
```
Comment (jrz371) [inline glTF-BinExporter/RhinoDocGltfConverter.cs:182]: Make this name fix it's own method as well

**Review by jrz371 — COMMENTED** [review]
Only these three small fixes.

If you can, please add a test model as well for verifying the layers and subobject

**Comment by jrz371** [comment]
#61 
#62

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 1 above)
Comment (pkratten) [reply glTF-BinExporter/RhinoDocGltfConverter.cs:282]: That is so the materials array and the meshes array match. I have a continue; somewhere else to deal with the null mesh. I am open to other solutions though!

**Comment by pkratten** [comment]
Here is a Scene to test the sub object and layer support:

[Sample.zip](https://github.com/Stykka/glTF-Bin/files/7020069/Sample.zip)

File: glTF-BinExporter/RhinoDocGltfConverter.cs (hunk 3 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoDocGltfConverter.cs]: 0fce1d63e4d043b2ac03b22223095392d8b69361

File: glTF-BinExporter/RhinoMaterialGltfConverter.cs (hunk 2 above)
Comment (jrz371) [reply glTF-BinExporter/RhinoMaterialGltfConverter.cs]: 0fce1d63e4d043b2ac03b22223095392d8b69361

**Comment by jrz371** [comment]
Great, thanks for the help!

**Pull request merged**